### PR TITLE
Allow each module to have its own class path

### LINKED ISSUE #10833: Annotation processors not run for hot reload of multi module maven project
**Describe the bug**
I have a multi module maven project with the following structure:

project:
- web (quarkus project)
- service (jar packaging)

The service modul has a provided dependency to an annotation processor - hibernate-jpamodelgen to be exact.
During a normal "mvn compile", the annotation processor is run, and the static metamodel gets generated.

Now, i started dev mode.
If I change one of my entity classes, the annotation processor is not run again during hot reload.

As a workaround, I can add all annotation processors to the "web" project. The are then run correctly during dev mode reload.

Alternatively, I could also change the ap to compile scope in the service project, which hints at class path problems.

**Expected behavior**
Compilation is done with the correct class path of a specific module. All annotation processors of a module are executed.

**Actual behavior**
Annotation processor is not run on dev mode hot reload.

**To Reproduce**
1. Download the reproducer project: 
[ap-test.zip](https://github.com/quarkusio/quarkus/files/4943358/ap-test.zip)
2. Execute `mvn clean compile quarkus:dev` on the main project (ap-test), and open http://localhost:8080/hello in your browser
3. You will see output like `[id, ID]`. These are all fields currently in the static metamodel of the ExampleEntity class. The metamodel gets generated by the annotion processor.
4. Open the ExampleEntity class, and add/remove attributes.
5. Reload the browser page.
6. No change to the displayed metamodel fields.

If you now restart dev mode (`mvn clean compile quarkus:dev`), the change in metamodel is shown.

If you change the hibernate-jpamodelgen to compile scope, it is correctly re run on hot reload.

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
Linux martin 5.4.0-40-generic #44-Ubuntu SMP Tue Jun 23 00:01:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
 - Output of `java -version`: 
openjdk 11.0.7 2020-04-14 LTS
OpenJDK Runtime Environment Zulu11.39+15-CA (build 11.0.7+10-LTS)
OpenJDK 64-Bit Server VM Zulu11.39+15-CA (build 11.0.7+10-LTS, mixed mode)
 - Quarkus version or git rev: 
1.6.0.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /home/martin/.sdkman/candidates/maven/current
Java version: 11.0.7, vendor: Azul Systems, Inc., runtime: /home/martin/.sdkman/candidates/java/11.0.7-zulu
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.4.0-40-generic", arch: "amd64", family: "unix"

### PR DESCRIPTION
**Pull request by stuartwdouglas**
Each module now contains a list of dependencies,
so individual class paths are respected.

Partially Fixes #10833, as this will now work when
quarkus:dev is run from the root, but not when it is
run from a sub folder. As we currently can't resolve
the dependencies of the other project.

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/dev/ClassLoaderCompiler.java [hunk 1]
```diff
@@ -1,200 +0,0 @@
...
-
-import io.quarkus.bootstrap.app.CuratedApplication;
-import io.quarkus.bootstrap.model.AppDependency;
-
-/**
- * Class that handles compilation of source files
- *
- * @author Stuart Douglas
- */
-public class ClassLoaderCompiler implements Closeable {
```
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/ClassLoaderCompiler.java:35]: This was renamed as the old name is no longer relevant, it used to compile by reading the current ClassPath from the ClassLoader, and it has not done that for a very very long time.

**Comment by gsmet** [comment]
Force pushed an update to fix formatting.

**Comment by gsmet** [comment]
This one is doomed. I rebased to fix an easily fixable conflict.

File: devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java [hunk 2]
```diff
@@ -460,7 +460,7 @@ private void addLocalProject(Project project, DevModeContext context, Set<AppArt
                 resourcesOutputPath,
                 sourceParentPaths,
                 project.getBuildDir().toPath().resolve("generated-sources").toAbsolutePath().toString(),
-                project.getBuildDir().toString());
+                project.getBuildDir().toString(), null);
```
Comment (gsmet) [inline devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:463]: I'm surprised we pass `null` here?

File: test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java [hunk 3]
```diff
@@ -284,7 +284,7 @@ private DevModeContext exportArchive(Path deploymentDir, Path testSourceDir, Pat
                             classes.toAbsolutePath().toString(), deploymentResourcePath.toAbsolutePath().toString(),
                             deploymentSourceParentPath.toAbsolutePath().toString(),
                             targetDir.resolve("generated-sources").toAbsolutePath().toString(),
-                            targetDir.toAbsolutePath().toString()));
+                            targetDir.toAbsolutePath().toString(), null));
```
Comment (gsmet) [inline test-framework/junit5-internal/src/main/java/io/quarkus/test/QuarkusDevModeTest.java:287]: Same here.

File: devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java (hunk 2 above)
Comment (gsmet) [reply devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java:463]: Meaning an empty list would probably be better if that's what we want to pass.

**Comment by gsmet** [comment]
Tests are failing:
```

**Comment by stuartwdouglas** [comment]
This breaks remote dev mode in a way that I don't really understand yet.

The idea with the nulls is that if no module specific class path is used the existing global CP (i.e. all dependencies of the current application) would be used instead, to allow us to implement this incrementally.

**Comment by gsmet** [comment]
@stuartwdouglas should we close this one?

**Comment by stuartwdouglas** [comment]
Closing for now

**Pull request closed without merging**